### ISSUE TITLE
Tab-completion apparently got slower from v7.1 to v7.9

### ISSUE THREAD

**Issue by augustogoulart** [ipython/ipython#11942]
Split from https://github.com/ipython/ipython/issues/11908

There’s a noticeable lag on the tab-completion for IPython on v7.9 that’s not present on v7.1.

🎥 Video recording for 7.9:
https://recordit.co/aKoZsujIHx

🎥 Video recording for 7.1:

**v7.2 seemed to be _a bit_ slower already**, so I'll run some tests to check whether or not something degraded the perceived performance of the tab-completion overtime.

**Comment by normanius** [comment]
It seems that the larger an object / module, the longer it takes to compile the list of suggestions. The lag is present for the first time only, successive attempts are fast.

For the pandas module, I measured a lag of a bit more than 2s when checking the first time, see screenshot.

```
import pandas as pd
pd.<TAB>
```
![image](https://user-images.githubusercontent.com/11406002/68855375-67ab3000-06de-11ea-8cac-54f52bb73ee4.png)

Note that the presentation of suggestions looks slightly different for the second tab-completion attempt. Not sure if this is in any way relevant for this issue, though as a user I'm a bit surprised that the result is not the same.

![image](https://user-images.githubusercontent.com/11406002/68855760-3e3ed400-06df-11ea-8b2f-96eaf860e1ad.png)

**Comment by normanius** [comment]
For comparison: this is the output for `pd.<TAB>` in IPython 7.1.0. There was no observable lag.

![image](https://user-images.githubusercontent.com/11406002/68856845-c1f9c000-06e1-11ea-963a-53ac152723ed.png)

**Comment by Carreau** [comment]
This is due to jedi; and the reason is when jedi inspect the module there is a non-reducible overhead that need to scan the target library once.

You can see in there:

https://github.com/ipython/ipython/blob/master/IPython/core/completer.py#L1852-L1882

That if it takes too long we try to stop as soon as we can; but _if_ we have enough time left; we try to compute the types and signature of all the functions we can. Each subsequent tab completion will thus (usually) manage to get the type signature of a few other functions, and hence why the tab completion can change after each keypress.

So the behavior is not too surprising; and most of the work should likely try to be in making jedi more performant; or more asynchronous so that the completions appear to be instantaneous.

Last time I check the bottleneck was loading and parsing the cache files from disk.

**Comment by normanius** [comment]
Thanks for looking into this. I see, we face a trade-off problem here: snappy behavior vs. more complete/better completions. 

But the current situation is really bad: For example, I don't get any suggestions for the numpy package, even after waiting for a long time (minutes). Likewise, for nested object completions, Jedi seems not to do a proper job.

As you suggest in #11945, disabling jedi fixes the problem for me:
```ipython --IPCompleter.use_jedi=False```

**Comment by ssanderson** [comment]
**TL;DR** with the combination of https://github.com/ipython/ipython/pull/12299 and using the head of master Jedi, completion performance seems to be significantly improved. I have a couple ideas for possible further improvements to completion performance, but they might not even be necessary.

---

I've been working yesterday and today on trying to understand and improve the performance of jedi-based completions in IPython. The first step toward this is making it easy to profile completion performance (c.f. https://github.com/ipython/ipython/pull/12298). With that in place, it's relatively easy to start looking into where things are slow.

I've gathered a few notes from looking at profiles. I've mostly been looking at the performance of `pd.<TAB>` and `pd.DataFrame.<TAB>` so far:

1. There are at least two distinct cases to think about when measuring performance of completions: first time completions and subsequent completions. These are important to separate because there are a few expensive operations that only happen the first time completions are requested. For example, parso (the parser library used by jedi) reads a grammar file the first time completions are requested.

2. There's currently an unnecessarily slow codepath that's unrelated to Jedi where we're initializing a big table of unicode character names when it's not needed. This should be fixed by https://github.com/ipython/ipython/pull/12299 (now merged). See that PR description for profiling details.

3. With the unicode issue out of the way, we're spending most of our completion time in Jedi's internals. However, there's some good news: the latest version of Jedi on GitHub seems to be significantly faster than the currently released PyPI version, at least for the completions I've been testing.

    On my machine, a first time completion of `pd.<TAB>` takes about .79 seconds with Jedi 0.17, and a second time completion takes about .77 seconds. Using head of Jedi's master branch, a first time completion of `pd.<TAB>` takes about .17 seconds (after applying https://github.com/ipython/ipython/pull/12299), and a second time completion takes about .044 seconds. The above times are with profiling enabled. With profiling disabled they drop down to 0.08 (barely noticeable for me) and 0.02 seconds (not noticeable)

    Qualitatively, there's a big difference between each of those three timings. .79 seconds is enough that completions feel noticeably sluggish. .17 seconds is much better, but still feels a bit "laggy". .044 seconds feels basically instant. It would be amazing if we could get all completions down to this speed (probably just getting down to under 100ms would be a big win).

    It's not totally clear to me where the speedup on head of master is coming from, but my best guess is https://github.com/davidhalter/jedi/commit/c314e1c36eb7debb9514cf96e7c501125f15782d based on the discussion in https://github.com/davidhalter/jedi/issues/1422. If that's indeed the case, then it seems like just releasing a new version of Jedi would be a big win for many IPython users.

4. I'm not totally confident in these conclusions because the probe effect of profiling seems to be pretty significant, but looking at the profile for first-time completions on latest jedi, it seems like the next biggest hotspot is two calls to `parso.load_grammar`, which take up 0.09 of the 0.17 seconds:

    ```
    .completion_profiles/pd.tab_first_time_latest_jedi% stats 25
    Sat May  9 11:55:04 2020    .completion_profiles/pd.tab_first_time_latest_jedi

             361737 function calls (349398 primitive calls) in 0.167 seconds

       Ordered by: cumulative time
       List reduced from 809 to 25 due to restriction <25>

       ncalls  tottime  percall  cumtime  percall filename:lineno(function)
          116    0.001    0.000    0.167    0.001 /home/ssanderson/projects/IPython/IPython/core/completer.py:1867(_completions)
    3795/1993    0.005    0.000    0.118    0.000 /home/ssanderson/projects/jedi/jedi/cache.py:103(wrapper)
            1    0.000    0.000    0.105    0.105 /home/ssanderson/projects/IPython/IPython/core/completer.py:1997(_complete)
            1    0.000    0.000    0.103    0.103 /home/ssanderson/projects/IPython/IPython/core/completer.py:1359(_jedi_matches)
            1    0.000    0.000    0.096    0.096 /home/ssanderson/projects/jedi/jedi/api/__init__.py:840(__init__)
            1    0.000    0.000    0.096    0.096 /home/ssanderson/projects/jedi/jedi/api/__init__.py:124(__init__)
            1    0.000    0.000    0.090    0.090 /home/ssanderson/projects/jedi/jedi/inference/__init__.py:85(__init__)
            2    0.000    0.000    0.090    0.045 /home/ssanderson/.virtualenvs/IPython/lib/python3.8/site-packages/parso/grammar.py:226(load_grammar)
            2    0.000    0.000    0.090    0.045 /home/ssanderson/.virtualenvs/IPython/lib/python3.8/site-packages/parso/grammar.py:234(load_grammar)
            2    0.000    0.000    0.089    0.045 /home/ssanderson/.virtualenvs/IPython/lib/python3.8/site-packages/parso/grammar.py:209(__init__)
            2    0.000    0.000    0.089    0.045 /home/ssanderson/.virtualenvs/IPython/lib/python3.8/site-packages/parso/grammar.py:32(__init__)
            2    0.001    0.000    0.089    0.045 /home/ssanderson/.virtualenvs/IPython/lib/python3.8/site-packages/parso/pgen2/generator.py:235(generate_grammar)
            1    0.000    0.000    0.068    0.068 /home/ssanderson/projects/jedi/jedi/api/environment.py:34(get_grammar)
          115    0.000    0.000    0.062    0.001 /home/ssanderson/projects/IPython/IPython/core/completer.py:1791(completions)
          115    0.000    0.000    0.055    0.000 /home/ssanderson/projects/jedi/jedi/api/classes.py:713(type)
    ```

    One thing that's surprising to me about this is that there are two `load_grammar` calls, rather than one. I wonder if we're (accidentally?) constructing the grammar twice.

    The other relatively hot spot seems to be `Completion._get_docstring_signature`, which is getting called during cache population from `Completion.type`. It looks to me like what's happening here is that Jedi populates a cache entry for completions, and that cache entry contains a type, a docstring, and a signature. IPython ends up using that signature later on in the case where `completion.type == 'function'`, but it only gets displayed when the user tabs through different possible completions. I think we might be able to speed up this code path by deferring more work so that types and signatures only get computed if they're needed.

@davidhalter if you're around, I'd be curious for your thoughts on whether the above notes on Jedi make sense and/or if there's anything obvious I'm missing here. I'd be interested in trying to work on performance improvements upstream if there's work that seems worthwhile.

Profiles I've been looking at for the curious: [completion_profiles.zip].(https://github.com/ipython/ipython/files/4604134/completion_profiles.zip)

**Comment by Carreau** [comment]
Thanks @ssanderson for looking into this !

> The other relatively hot spot seems to be Completion._get_docstring_signature, which is getting called during cache population from Completion.type. It looks to me like what's happening here is that Jedi populates a cache entry for completions, and that cache entry contains a type, a docstring, and a signature. IPython ends up using that signature later on in the case where completion.type == 'function', but it only gets displayed when the user tabs through different possible completions. I think we might be able to speed up this code path by deferring more work so that types and signatures only get computed if they're needed.

Your profiling might be highly biased by how IPython ask jedi for completion. we do ask jedi for `_get_docstring_signature` while iterating over the completions and stop if the "tab completion time budget" is over. 

You may want to set this budget to either 0, or infinite when profiling.

Another possibility to think about is to look at making completions UI lazy in prompt toolkit (Maybe it is already?), and actually ask for the type signature and co when we actually want to display it (or maybe cache a screen or so and lazily load as the user looks around.

**Comment by ssanderson** [comment]
> Your profiling might be highly biased by how IPython ask jedi for completion. we do ask jedi for _get_docstring_signature while iterating over the completions and stop if the "tab completion time budget" is over.

We're definitely hitting the timeout on the first completion when using using jedi 0.17, since the default timeout is 400ms and the completions are taking around 700ms to produce for me.

When using the latest jedi from GitHub, the slowest timing I'm seeing is around 170ms, so I don't think we're hitting that timeout anymore. I just ran with `jedi_compute_type_timeout` set to 0 and I'm not seeing a difference in profiling when using latest jedi.

> Another possibility to think about is to look at making completions UI lazy in prompt toolkit (Maybe it is already?), and actually ask for the type signature and co when we actually want to display it

I think this might already be possible based on https://python-prompt-toolkit.readthedocs.io/en/master/pages/reference.html?highlight=__pt_formatted_text__#module-prompt_toolkit.formatted_text. It looks like most places where you can pass text to Prompt Toolkit, you can also pass any object that implements a `__pt_formatted_text__` method. That probably means we can just add a class that wraps a Jedi `Completion` object and only requests type information if `__pt_formatted_text__` gets called. Depending on how prompt toolkit draws the completions dropdown, that might prevent us from requesting type information for things that won't be displayed. I'll take a look and see if I can get a prototype of this working.

**Comment by davidhalter** [comment]
> If that's indeed the case, then it seems like just releasing a new version of Jedi would be a big win for many IPython users.

A new release is coming soon I guess. There's just a few things around generics that I want to clean up.

> @davidhalter if you're around, I'd be curious for your thoughts on whether the above notes on Jedi make sense and/or if there's anything obvious I'm missing here. I'd be interested in trying to work on performance improvements upstream if there's work that seems worthwhile.

You're pretty much 100% correct. Thanks a lot for the in-depth investigation!

> The other relatively hot spot seems to be Completion._get_docstring_signature

If it's possible to do that lazy, please go for it. It doesn't make sense at all to get the docstrings/signatures for 500 objects if you only display one at a time.

> the next biggest hotspot is two calls to parso.load_grammar, which take up 0.09 of the 0.17 seconds:

This could probably be optimized by pickling the Python grammars in parso. I'm not a big fan of that for various reasons, but I'm open to discuss it.

> One thing that's surprising to me about this is that there are two load_grammar calls, rather than one. I wonder if we're (accidentally?) constructing the grammar twice.

No this is intentional. To avoid dealing with stubs in different grammars, we just use the same grammar all the time. So I guess in your Python version two grammars are loaded, while in `3.7` it would only load one grammar.
